streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -250,127 +237,137 @@ abstract class ZSink[-R, +E, -I, +Z] private (
    * and emits the sink's results as outputs. The sink will be restarted when
    * it ends.
    */
-  def toTransducer: ZTransducer[R, E, I, Z] =
+  def toTransducer(implicit ev: L <:< I): ZTransducer[R, E, I, Z] =
     ZTransducer {
       ZSink.Push.restartable(push).map {
-        case (push, restart) =>
-          (input: Option[Chunk[I]]) =>
-            push(input).foldM(
-              {
-                case Left(e)  => ZIO.fail(e)
-                case Right(z) => restart.as(Chunk.single(z))
-              },
-              _ => UIO.succeedNow(Chunk.empty)
-            )
+        {
+          case (push, restart) => {
+
+            def go(input: Option[Chunk[I]]): ZIO[R, E, Chunk[Z]] =
+              push(input).foldM(
+                {
+                  case (Left(e), _) => ZIO.fail(e)
+                  case (Right(z), leftover) =>
+                    restart *> {
+                      if (leftover.isEmpty || input.isEmpty) {
+                        ZIO.succeed(Chunk.single(z))
+                      } else {
+                        go(Some(leftover).asInstanceOf[Option[Chunk[I]]]).map(more => Chunk.single(z) ++ more)
```

Comment:
Why is the cast necessary here?

Reply:
`leftover` has type `Chunk[L]`, not `Chunk[I]`. `<:<` is not automatically applied.